README.markdown
TRAMMI
======
[![NPM version](https://badge.fury.io/js/trammi.png)](http://badge.fury.io/js/trammi)
[![NGN Dependencies](https://david-dm.org/albertosantini/node-trammi.png)](https://david-dm.org/albertosantini/node-trammi)

TramMI is a web app to estimate the time of arrival of tram and bus, the public
transport of Milan city.

The official apps display only the latest estimated time of arrival.

After you selected the number line and the stop, it is displayed immediately the
waiting time of the next arrival. Then, every minute, the estimated time of
arrivals of the next vehicles are displayed.

Stack software
==============

Frontend
--------

- [Pure](http://purecss.io/)
- [Lato typeface](http://www.google.com/webfonts/specimen/Lato)
- [Yahoo! User Interface](http://yuilibrary.com/)

Browser Support and Testing
---------------------------

* IE 7+
* Latest Stable: Firefox, Chrome, Safari
* iOS 6.x
* Android 4.x

Backend
-------

- [Node.js](http://nodejs.org/)

Dataset
-------

Open data to build tram stops database:

- [Servizio di trasporto pubblico locale dal 16/05/2013 al 05/07/2013](http://dati.comune.milano.it/dato/item/180-180-programmato-del-servizio-di-trasporto-pubblico-locale-dal-16052013-al-05072013.html)
- [Servizio di trasporto pubblico locale dal 22/06/2013 al 30/06/2013](http://dati.comune.milano.it/dato/item/184-184-programmato-del-servizio-di-trasporto-pubblico-locale-dal-22062013-al-30062013.html)
- [Servizio di trasporto pubblico locale dal 01/07/2013 al 01/08/2013](http://dati.comune.milano.it/dato/item/185-185-programmato-del-servizio-di-trasporto-pubblico-locale-dal-01072013-al-01082013.html)
- [Servizio di trasporto pubblico locale dal 02/08/2013 al 01/09/2013](http://dati.comune.milano.it/dato/item/189-189-programmato-del-servizio-di-trasporto-pubblico-locale-dal-02082013-al-01092013.html)

API example
===========

    var trammi = require("trammi");

    // stop code 12254, line 1
    trammi.getStopInfo(12254, 1, function (err, info) {
        console.log(info);
    });

    // line 1, stop number 39
    trammi.getWaitingTime("1a", 39, function (err, timetable) {
        var i, n;

        n = timetable.length;
        for (i = 0; i < n; i += 1) {
            console.log("Vehicle", i + 1, "Waiting =", timetable[i]);
        }
    });

    trammi.startServer(8081);

Installation
============

To install with [npm](http://github.com/isaacs/npm):

    npm install trammi

Tested with node 0.10.x.

Notes
=====

At the moment only tram lines are implemented.

In support folder, there is the script to recreate the database files, generated
in lib folder as 'trammiDB...' files.

Todo
====

Features
--------

- Total time for a path.

Business ideas
--------------

- Box office on Demand.

Weather info
------------

- [wunderground](http://www.wunderground.com/weather/api/d/documentation.html)
- [Historical weather data](http://allthingsr.blogspot.it/2012/04/getting-historical-weather-data-in-r.html)
- [Tempo di Milano](http://www.tempodimilano.it/)

API Backend
===========

getStops(query, cb)
--------

It finds a list of stops if the words, separated by space(s), contained in the
query, match the field 'acdescr' of the stop.

**Params**

- *query* the chars typed in the autocomplete

**Callback response**

- *stops* array of stops matching the query.


getStopInfo(stop, stopLine, cb)
-----------

It retrieves the info of the stop.

**Params**

- *stop* code of the stop.

- *stopLine*  number of the line.

**Callback response**

- *stopInfo* array.

    - *line* number of the line.

    - *eta* estimated time of arrival (string).

getStopsList(lineName, callback)
------------

It retrieves all the stops of a line. On server side there is a cache with a
60 secs. expiring time.

**Params**

- *lineName* name of the line.

**Callback response**

- *stopInfo* array

    - *descr* description of the top.

    - *eta* estimated time of arrival (number).

getWaitingTime(lineName, stopNumber, cb)
--------------

It retrieves the arrivals of all the vehicles until a stop.

**Params**

- *lineName* name of the line.

- *stopNumber* index of the stop related to the line.

**Callback response**

- *timetable* array containing the estimated time of arrivals in minutes.


startServer(port, publicDir, appInst, expressInst)
-----------

It starts a server providing a frontend for the module.

The method adds the following routes to the express instance:

- /trammi/getStops

- /trammi/getStopInfo

- /trammi/getTimetable

**Params**

- *port* port of the server.

- *publicDir* directory of public folder containing the frontend.

- *appInst* an instance of the express.HTTPServer, created with createServer().

- *expressInst* an instance of the express module.




HISTORY.markdown
1.1.1 / 2013-09-12
==================

* Fix clear icon.
* Update to Pure 0.3.0.
* Update express 3.4.0.

1.1.0 / 2013-09-06
==================

* Add stop code in the search.
* Add clear icon in the search.
* Replace jQuery module with cheerio one.
* Update to YUI 3.12.0.

1.0.0 / 2013-08-26
==================

* Update open data links.

0.9.1 / 2013-07-20
==================

* Add rounded stop search.
* Update to YUI 3.11.0.
* Update to Pure 0.2.1.
* Update dev dep xml2js 0.2.8.

0.9.0 / 2013-07-13
==================

* Update to Express 3.x.
* Review README text including stack software info.
* Add npm version and dependency badges.
* Refactor jshint settings.

0.7.0 / 2013-06-16
==================

* Use Pure css framework.
* Use Lato font.
* Update to YUI 3.10.3.

0.6.8 / 2013-06-06
==================

* Update to YUI 3.10.2.

0.6.7 / 2013-05-30
==================

* Add transition in the timetable.
* Fix the input size on iphone.
* Review the main text.

0.6.6 / 2013-05-29
==================

* Add timedate in the timetable.

0.6.5 / 2013-05-27
==================

* Fix again the stop request.

0.6.4 / 2013-05-27
==================

* Update to YUI 3.10.1.
* Fix an issue with stop request due to a change in ATM backend.

0.6.3 / 2013-05-05
==================

* Update to YUI 3.10.0.
* Fix soppressa and notturna eta.

0.6.2 / 2013-04-03
==================

* Update to YUI 3.9.1.
* Update to Bootstrap 2.3.1.
* Update icons.

0.6.1 / 2013-01-12
==================

* Replaced line stop image with a tram one.
* Moved the timetable view to the center below the search field.

0.6.0 / 2012-12-26
==================

* Update to YUI 3.8.0.
* Update to Bootstrap 2.2.2.
* Refactor layout using a clean stop view based on a list of stops (vs. button).

0.5.8 / 2012-11-09
==================

* Clean stops in timetable.

0.5.7 / 2012-11-08
==================

* Fix font and break in timetable.
* Update to Bootstrap 2.2.1.

0.5.6 / 2012-10-30
==================

* Update to Bootstrap 2.2.0, YUI 3.7.3.

0.5.5 / 2012-09-19
==================

* Fix the font size for the autocomplete box.

0.5.4 / 2012-09-17
==================

* Update to Bootstrap 2.1.1.

0.5.3 / 2012-03-31
==================

* Added a few logs in startServer.

0.5.2 / 2012-03-29
==================

* Fixed joyent setup.

0.5.1 / 2012-03-29
==================

* Moved frontend folder coherent with joyent setup.
* Moved db creation to support folder.

0.5.0 / 2012-03-28
==================

* Changed dnode with jsonp.

0.4.11 / 2012-03-15
===================

* Updated tw bootstrap 2.0.2.
* Deleted connect logger.
* Fix history doc.

0.4.10 / 2012-03-04
===================

* Added connect logger.

0.4.9 / 2012-03-03
==================

* Rollback to the original approach.

0.4.8 / 2012-03-03
==================

* Replaced express module with connect one and moved connect instance outside
startServer method.

0.4.7 / 2012-03-03
==================

* Replaced connect module with express one.

0.4.6 / 2012-02-28
==================

* Removed the forced jsdom@0.2.10 because released jsdom@0.2.13.

0.4.5 / 2012-02-27
==================

* Forced again jsdom@0.2.10 due to an incompaatibility with nodester 0.4.9.

0.4.4 / 2012-02-27
==================

* Removed the direct dependency for jsdom@0.2.10.

0.4.3 / 2012-02-25
==================

* Forcing jsdom@0.2.10 because contextify@0.1 is not supported on windows.

0.4.2 / 2012-02-25
==================

* Updated tw bootstrap 2.0.1.
* The timetable is called after the stop info and every minute.

0.4.1 / 2012-02-04
==================

* Moved conditional IE scripts in yuiConfig.
* Replaced search with indexOf in getStops.

0.4.0 / 2012-02-03
==================

* Added cache on server side for the timetable and the autocomplete results.
* Moved autocompletion search on server side.
* Compatibility with IE 7/8/9.
* Used the final bootstrap css file.

0.3.0 / 2012-01-30
==================

* New layout for timetable.

0.2.0 / 2012-01-25
==================

* New layout using twitter bootstrap css.

0.1.0 / 2012-01-24
==================

* Initial release: only tram lines are implemented.



lib/trammi.js
"use strict";

var cache = require("memory-cache"),
    request = require("request"),
    cheerio = require("cheerio"),
    lines = require("./trammiDBLines"),
    stopsAC = require("./trammiDBStops");

function getStopInfo(stop, stopLine, cb) {
    request({
        uri: "http://www.atm.it/gmtools/xmllvstop.asp",
        method: "POST",
        body: "<stoprequest>" +
            "<code>" + stop + "</code>" +
            "<linename>" + stopLine + "</linename>" +
            "</stoprequest>",
        headers: {
            Cookie: "ASP.NET_SessionId=tkrqqmnt5fddm32hphrmgdrq; ",
            "User-Agent": "Chrome/27.0.1453.94 Safari/537.36",
            "Referer": "http://www.atm.it/it/Giromilano/Pagine/default.aspx"
        }
    }, function (err, res, body) {
        var $,
            stopInfo = [];

        if (err) {
            console.log("getStopInfo request error.");
            cb(true);

            return;
        }

        $ = cheerio.load(body.replace(/<!\[CDATA\[([^\]]+)]\]>/ig, "$1"));

        $(".tblttinfo > tr").slice(1).each(function () {
            stopInfo.push({
                line: $("td:nth-of-type(1)", this).text(),
                eta: $("td:nth-of-type(2)", this).text()
            });
        });

        if (stopInfo.length > 0) {
            cb(false, stopInfo);
        } else {
            cb(true);
        }

    });
}
exports.getStopInfo = getStopInfo;

function getStopsList(l, callback) {
    var line = lines[l].line,
        stops = lines[l].info.stop,
        n = stops.length,
        data = [],
        counter = 0,
        i;

    function handleStopInfo(stop, index, cb) {
        var s = stop.$;

        getStopInfo(s.scode, line, function (err, stopInfo) {
            var line = lines[l].line, stopEta, stopLine, m, k;


            if (!err) {
                m = stopInfo.length;
                for (k = 0; k < m; k += 1) {
                    stopLine = stopInfo[k].line;

                    if (stopLine === line) {
                        stopEta = stopInfo[k].eta;

                        if (stopEta === "") {
                            stopEta = "1 min";
                        } else if (stopEta === "in arrivo") {
                            stopEta = "1 min";
                        } else if (stopEta === "partenza") {
                            stopEta = "1 min";
                        } else if (stopEta === "--") {
                            stopEta = "1 min";
                        } else if (stopEta === "ricalcolo") {
                            stopEta = "1 min";
                        } else if (stopEta === "in aggiorn") {
                            stopEta = "5 min";
                        } else if (stopEta === "soppressa") {
                            stopEta = "60 min";
                        } else if (stopEta === "notturna") {
                            stopEta = "60 min";
                        } else if (stopEta === "coll bus") {
                            stopEta = "60 min";
                        } else if (stopEta === "*") {
                            stopEta = "60 min";
                        } else {
                            stopEta = stopInfo[k].eta;
                        }

                        cb(false, index, {
                            descr: s.sdescr,
                            eta: stopEta
                        });
                    }
                }
            } else {
                cb(true);
            }
        });
    }

    function next(err, index, res) {
        counter += 1;

        if (!err) {
            data[index] = res;
        }

        if (counter === n) {
            callback(false, data);
        }
    }

    for (i = 0; i < n; i += 1) {
        handleStopInfo(stops[i], i, next);
    }
}
exports.getStopsList = getStopsList;

function getWaitingTime(lineName, stopNumber, cb) {

    function getPreviousStop(index, s) {
        var k;

        for (k = index - 1; k > 0; k -= 1) {
            if (s[k]) {
                break;
            }
        }

        return k;
    }

    getStopsList(lineName, function (err, stops) {
        var i, p, n, eta, eta2, t = 0, last, total = [{eta: 0, pnum: -1}];
        // console.log(require("util").inspect(stops, false, null));

        if (!err) {
            // console.log(lines[lineName].way);

            n = stops.length;
            for (i = stopNumber - 1; i > 0; i -= 1) {
                if (stops[i]) {
                    if (!stops[i - 1]) {
                        p = getPreviousStop(i, stops);
                    } else {
                        p = i - 1;
                    }

                    eta = parseInt(stops[i].eta, 10);
                    eta2 = 0;
                    if (stops[p]) {
                        eta2 = parseInt(stops[p].eta, 10);
                    }

                    // console.log(i, "@", stops[i].descr, "=", eta);

                    if (eta >= eta2 && t === 0) {
                        // console.log("a");
                        t = eta;
                    } else if (eta === eta2) {
                        // console.log("b");
                        t += 1;
                    } else if (eta < eta2) {
                        // console.log("c");
                        last = total.pop();
                        total.push(last);
                        if (t === 0) {
                            total.push({
                                eta: last.eta + eta,
                                pnum: i + 1
                            });
                        }

                        if (t !== 0) {
                            total.push({
                                eta: last.eta + t,
                                pnum: i + 1
                            });
                            t = 0;
                        }
                    } else if (p === 0) {
                        // console.log("d");
                        last = total.pop();
                        total.push(last);
                        total.push({
                            eta: last.eta + (t !== 0 ? t : eta),
                            pnum: i + 1
                        });
                    }
                }
            }
            if (total.length === 1) {
                if (t === 0) {
                    eta = stops[i].eta;
                } else {
                    eta = t;
                }

                total.push({
                    eta: eta,
                    pnum: i + 1
                });
            }
            // console.log(i, "@", stops[i].descr, "=", eta);

            total.shift();
            cb(false, total);
        }
    });
}
exports.getWaitingTime = getWaitingTime;

function startServer(port, directory, appInst, expressInst) {
    var express = expressInst || require("express"),
        path = require("path"),
        p = port || 8081,
        d = path.resolve(__dirname, directory || __dirname + "/public"),
        app = appInst || express();

    console.log("trammi static folder is " + d);

    app.configure(function () {
        // app.use(express.favicon(d + '/favicon.ico'));
        app.use(express["static"](d));
        app.use(express.bodyParser());
        app.use(express.errorHandler());
        app.use(express.cookieParser());
    });

    app.get("/trammi/getStops", function (req, res) {
        var query = req.query.q,
            cacheKey = "queryCache#" + query,
            words = query.toLowerCase().split(" "),
            acdescr,
            i,
            j,
            n = stopsAC.length,
            m = words.length,
            stops,
            found;

        stops = cache.get(query);

        if (!stops) {
            stops = [];

            for (i = 0; i < n; i += 1) {
                found = true;
                for (j = 0; j < m; j += 1) {
                    acdescr = stopsAC[i].acdescr.toLowerCase();
                    if (acdescr.indexOf(words[j]) === -1) {
                        found = false;
                        break;
                    }
                }
                if (found) {
                    stops.push(stopsAC[i]);
                }
            }

            cache.put(cacheKey, stops);
        }

        res.jsonp(stops);
    });

    app.get("/trammi/getStopInfo", function (req, res) {
        var stopNumber = req.query.stopNumber,
            lineCode = req.query.lineCode;

        getStopInfo(stopNumber, lineCode, function (err, stopInfo) {
            var i, n, stopLine;

            if (!err) {
                n = stopInfo.length;

                for (i = 0; i < n; i += 1) {
                    stopLine = stopInfo[i].line;

                    if (stopLine === lineCode) {
                        res.jsonp(stopInfo);
                    }
                }
            }

        });
    });

    app.get("/trammi/getTimetable", function (req, res) {
        var lineName = req.query.lineName,
            stopNumber = req.query.stopNumber,
            cacheKey = "stopCache#" + lineName + "/" + stopNumber,
            timetable;

        timetable = cache.get(cacheKey);

        if (!timetable) {
            getWaitingTime(lineName, stopNumber, function (err, timetable) {
                if (!err) {
                    cache.put(cacheKey, timetable, 60000);
                    res.jsonp(timetable);
                }
            });
        } else {
            res.jsonp(timetable);
        }
    });

    if (!appInst) {
        app.listen(p);
        console.log("trammi listening on port " + p);
    }
}
exports.startServer = startServer;

